// mongodb-queue/mongodb-queue.js
/**
 *
 * mongodb-queue.js - Use your existing MongoDB as a local queue.
 *
 * Copyright (c) 2014 Andrew Chilton
 * - http://chilts.org/
 * - andychilton@gmail.com
 *
 * License: http://chilts.mit-license.org/2014/
 *
 **/

var crypto = require('crypto')

// some helper functions
function id() {
    return crypto.randomBytes(16).toString('hex')
}

function now() {
    return (new Date()).toISOString()
}

function nowPlusSecs(secs) {
    return (new Date(Date.now() + secs * 1000)).toISOString()
}

module.exports = function(mongoDbClient, name, opts) {
    return new Queue(mongoDbClient, name, opts)
}

// the Queue object itself
function Queue(mongoDbClient, name, opts) {
    if ( !mongoDbClient ) {
        throw new Error("mongodb-queue: provide a mongodb.MongoClient")
    }
    if ( !name ) {
        throw new Error("mongodb-queue: provide a queue name")
    }
    opts = opts || {}

    this.name = name
    this.col = mongoDbClient.collection(name)
    this.visibility = opts.visibility || 30
    this.delay = opts.delay || 0

    if ( opts.deadQueue ) {
        this.deadQueue = opts.deadQueue
        this.maxRetries = opts.maxRetries || 5
    }
}

Queue.prototype.createIndexes = function(callback) {
    var self = this

    self.col.createIndex({ deleted : 1, visible : 1 }, function(err, indexname) {
        if (err) return callback(err)
        self.col.createIndex({ ack : 1 }, { unique : true, sparse : true }, function(err) {
            if (err) return callback(err)
            callback(null, indexname)
        })
    })
}

Queue.prototype.add = function(payload, opts, callback) {
    var self = this
    if ( !callback ) {
        callback = opts
        opts = {}
    }
    var delay = opts.delay || self.delay
    var visible = delay ? nowPlusSecs(delay) : now()

    var msgs = []
    if (payload instanceof Array) {
        if (payload.length === 0) {
            var errMsg = 'Queue.add(): Array payload length must be greater than 0'
            return callback(new Error(errMsg))
        }
        payload.forEach(function(payload) {
            msgs.push({
                visible  : visible,
                clientReference:payload.clientReference,
                appid:payload.appid,
                toCompanyid:payload.toCompanyid,
                companyid:payload.companyid,
                toAppid:payload.toAppid,
                ttl:payload.ttl || 30,
                priority:payload.priority ||  0,//优先级 0默认
                result:payload.result || "success",
                creatTime:now(),
                state:payload.state || 0,
                payload  : payload.payload ,
            })
        })
    } else {
        msgs.push({
            visible  : visible,
            clientReference:payload.clientReference,
            appid:payload.appid,
            companyid:payload.companyid,
            toCompanyid:payload.toCompanyid,
            toAppid:payload.toAppid,
            ttl:payload.ttl || 30,
            priority:payload.priority ||  0,//优先级 0默认
            result:payload.result || "success",
            creatTime:now(),
            state:payload.state || 0,
            payload  :payload.payload ,
        })
    }

    self.col.insertMany(msgs, function(err, results) {
        if (err) return callback(err)
        if (payload instanceof Array) return callback(null, '' + results.insertedIds)
        callback(null, '' + results.ops[0]._id)
    })
}

Queue.prototype.get = function(opts, callback) {
    var self = this
    if ( !callback ) {
        callback = opts
        opts = {}
    }

    var visibility = opts.visibility || self.visibility
    //添加搜索条件
    var clientReference=opts.clientReference;
    var query;
    if(clientReference){
         query = {
            deleted : null,
            visible : { $lte : now() },
            state:{$ne:2},//不等于2
            clientReference:clientReference//根据用户标识搜索
        }
    }else{
         query = {
            deleted : null,
            visible : { $lte : now() },
            state:{$ne:2}
            //  state:0,//只能get state=0的数据
        }
    }
/*    var query = {
        deleted : null,
        visible : { $lte : now() },
        state:{$ne:2}
      //  state:0,//只能get state=0的数据
    }*/
    var sort = {
        creatTime : 1
    }
    var update = {
        $inc : { tries : 1 },
        $set : {
            ack     : id(),
            visible : nowPlusSecs(visibility),
            state:1,//修改状态为已获取
        }
    }

    self.col.findOneAndUpdate(query, update, { sort: sort, returnOriginal : false }, function(err, result) {
        if (err) return callback(err)
        var msg = result.value
        if (!msg) return callback()

        // convert to an external representation
        msg = {
            // convert '_id' to an 'id' string
            id      : '' + msg._id,
            clientReference      : '' + msg.clientReference,
            ack     : msg.ack,
            companyid:msg.companyid,
            payload : msg.payload,
            creatTime:msg.creatTime,
            priority:msg.priority,
            tries   : msg.tries,
        }
        // if we have a deadQueue, then check the tries, else don't
        if ( self.deadQueue ) {
            // check the tries
            if ( msg.tries > self.maxRetries ) {
                // So:
                // 1) add this message to the deadQueue
                // 2) ack this message from the regular queue
                // 3) call ourself to return a new message (if exists)
                self.deadQueue.add(msg, function(err) {
                    if (err) return callback(err)
                    self.ack(msg.ack, function(err) {
                        if (err) return callback(err)
                        self.get(callback)
                    })
                })
                return
            }
        }

        callback(null, msg)
    })
}

Queue.prototype.ping = function(ack, opts, callback) {
    var self = this
    if ( !callback ) {
        callback = opts
        opts = {}
    }

    var visibility = opts.visibility || self.visibility
    var query = {
        ack     : ack,
        visible : { $gt : now() },
        deleted : null,
    }
    var update = {
        $set : {
            visible : nowPlusSecs(visibility)
        }
    }
    self.col.findOneAndUpdate(query, update, { returnOriginal : false }, function(err, msg, blah) {
        if (err) return callback(err)
        if ( !msg.value ) {
            return callback(new Error("Queue.ping(): Unidentified ack  : " + ack))
        }
        callback(null, '' + msg.value._id)
    })
}

Queue.prototype.ack = function(ack, callback) {
    var self = this
    console.log("now:"+now());
    var query = {
        ack     : ack,
        visible : { $gt : now() },//大于当前时间
        deleted : null,
    }
    var update = {
        $set : {
            deleted : now(),
            state:2,//修改状态为已处理
        }
    }
    self.col.findOneAndUpdate(query, update, { returnOriginal : false }, function(err, msg, blah) {
        if (err) return callback(err)
        if ( !msg.value ) {
            return callback(new Error("Queue.ack(): Unidentified ack : " + ack))
        }
        callback(null, '' + msg.value._id)
    })
}

Queue.prototype.clean = function(callback) {
    var self = this

    var query = {
        deleted : { $exists : true },
    }

    self.col.deleteMany(query, callback)
}

Queue.prototype.total = function(callback) {
    var self = this

    self.col.count(function(err, count) {
        if (err) return callback(err)
        callback(null, count)
    })
}

Queue.prototype.size = function(callback) {
    var self = this

    var query = {
        deleted : null,
        visible : { $lte : now() },
    }

    self.col.count(query, function(err, count) {
        if (err) return callback(err)
        callback(null, count)
    })
}

Queue.prototype.inFlight = function(callback) {
    var self = this

    var query = {
        ack     : { $exists : true },
        visible : { $gt : now() },
        deleted : null,
    }

    self.col.count(query, function(err, count) {
        if (err) return callback(err)
        callback(null, count)
    })
}

Queue.prototype.done = function(callback) {
    var self = this

    var query = {
        deleted : { $exists : true },
    }

    self.col.count(query, function(err, count) {
        if (err) return callback(err)
        callback(null, count)
    })
}

//by Gamehu 获取对应的queues

Queue.prototype.getQueuesByMyself = function(params,callback) {
    var self = this

    var query =params;
//   var fields={ sort: sort, returnOriginal : false };
    //建立索引
   // self.col.ensureIndex({"clientReference":1});//索引
    self.col.createIndex({"clientReference":1});
    self.col.createIndex({"clientReference":1,appid:1,companyid:1});//组合索引
    self.col.find(query).toArray(function(err,docs){
        console.log('find');
        if (err) {
            callback(err);
        }else{
            callback(null, docs);
        }

    });
}

//by Gamehu 获取对应的queue by cid

Queue.prototype.getQueueByCid = function(params,callback) {
    var self = this
    var query =params;
//   var fields={ sort: sort, returnOriginal : false };
    //建立索引
    self.col.ensureIndex({"clientReference":1});//索引
    self.col.findOne(query,function(err,doc) {
        if(err){
            callback(err);
        }else{
            callback(null,doc);
        }

    });
}
